syntax = "proto3";

package llmmcp.v1;

option go_package = "llm-mcp/core/internal/pb;pb";

service Core {
  rpc SubmitJob (SubmitJobRequest) returns (SubmitJobResponse);
  rpc GetJob (GetJobRequest) returns (GetJobResponse);
  rpc StreamJob (StreamJobRequest) returns (stream JobEvent);

  rpc RegisterWorker (RegisterWorkerRequest) returns (RegisterWorkerResponse);
  rpc ClaimJob (ClaimJobRequest) returns (ClaimJobResponse);
  rpc Heartbeat (HeartbeatRequest) returns (HeartbeatResponse);
  rpc CompleteJob (CompleteJobRequest) returns (CompleteJobResponse);
  rpc FailJob (FailJobRequest) returns (FailJobResponse);
  rpc ReportMetrics (ReportMetricsRequest) returns (ReportMetricsResponse);
  rpc ReportBenchmark (ReportBenchmarkRequest) returns (ReportBenchmarkResponse);
}

message SubmitJobRequest {
  string kind = 1;
  string payload_json = 2;
  int32 priority = 3;
  string source = 4;
  int32 max_attempts = 5;
  string deadline_at = 6;
}

message SubmitJobResponse {
  string job_id = 1;
}

message GetJobRequest {
  string job_id = 1;
}

message GetJobResponse {
  Job job = 1;
}

message StreamJobRequest {
  string job_id = 1;
}

message JobEvent {
  string job_id = 1;
  string type = 2;
  string message = 3;
  string ts = 4;
  string data_json = 5;
}

message RegisterWorkerRequest {
  WorkerInfo worker = 1;
}

message RegisterWorkerResponse {
  string worker_id = 1;
}

message ClaimJobRequest {
  string worker_id = 1;
  repeated string kinds = 2;
  int32 lease_seconds = 3;
}

message ClaimJobResponse {
  Job job = 1;
}

message HeartbeatRequest {
  string worker_id = 1;
  string job_id = 2;
  int32 extend_seconds = 3;
}

message HeartbeatResponse {
  bool ok = 1;
}

message CompleteJobRequest {
  string worker_id = 1;
  string job_id = 2;
  string result_json = 3;
  string metrics_json = 4;
}

message CompleteJobResponse {
  bool ok = 1;
}

message FailJobRequest {
  string worker_id = 1;
  string job_id = 2;
  string error = 3;
  string metrics_json = 4;
}

message FailJobResponse {
  bool ok = 1;
}

message ReportMetricsRequest {
  WorkerInfo worker = 1;
  string metrics_json = 2;
}

message ReportMetricsResponse {
  bool ok = 1;
}

message ReportBenchmarkRequest {
  Benchmark benchmark = 1;
}

message ReportBenchmarkResponse {
  bool ok = 1;
}

message Benchmark {
  string device_id = 1;
  string model_id = 2;
  string task_type = 3;
  int32 tokens_in = 4;
  int32 tokens_out = 5;
  int32 latency_ms = 6;
  double tps = 7;
  string meta_json = 8;
}

message Job {
  string id = 1;
  string kind = 2;
  string payload_json = 3;
  string status = 4;
  int32 attempts = 5;
  int32 max_attempts = 6;
  string lease_until = 7;
  string deadline_at = 8;
  string result_json = 9;
  string error = 10;
  int32 priority = 11;
  string queued_at = 12;
  string updated_at = 13;
}

message WorkerInfo {
  string id = 1;
  string name = 2;
  string platform = 3;
  string arch = 4;
  string host = 5;
  string tags_json = 6;
}
